build: add comparison of MEI versions

musicEnfanthen opened the pull request:
This PR adds an Ant target to run the MEI version comparison from build.xml. This prepares to include the comparison in the CI build process and gives the ability to include the comparison output file with releases and commits (for the dev branch).
By default, the target compares the current version (dev) to the previous version (MEI 4.0.1), but any other version can be compared via customized inputs.

New ant target tested locally and with docker.

Closes #1256

musicEnfanthen committed:
feat(build): add target for comparison of versions

musicEnfanthen committed:
docs(build): add new target to table of targets

musicEnfanthen committed:
utils(build): parameterize output folder and old.version.filename of comparison.xslt

bwbohl commented:
many thanks @musicEnfanthen, this looks all good but maybe we should accompany this with adding the compiled source to  https://github.com/music-encoding/schema/tree/main/4.0.1?

musicEnfanthen commented:
Yes, agree, at least the canonicalized source for 4.0.1 that is needed for the comparison. I also created "semi-manually" a canonicalized source for v3 (at least it gives the same results as the comparison that is online for v3 vs. v4)

musicEnfanthen committed:
Merge branch 'develop' into add-comparison

musicEnfanthen commented on build.xml:
```diff
@@ -390,7 +390,50 @@
+        <echo>Comparison done. See ${output.folder}/comparison.html for results.</echo>
+    </target>
+
     <target name="dist" depends="init-mei-classpath, canonicalize-source">
+
         <classfileset refid="mei.classpath"/>
```
```suggestion
    <target name="dist" depends="init-mei-classpath, canonicalize-source">
        <classfileset refid="mei.classpath"/>
```

musicEnfanthen committed:
fix(build.xml): remove empty line

musicEnfanthen commented:
Accompanying PR on schema repo is available: https://github.com/music-encoding/schema/pull/7

musicEnfanthen commented on build.xml:
```diff
@@ -390,6 +390,48 @@
+        <echo>Using source.file: ${source.file}</echo>
+
+        <!-- set old file to compare with (relative to comparison.xml) -->
+        <property name="old.default" value="MEI_v4.0.1_canonicalized.xml"/>
```
```suggestion
        <property name="old.default" value="mei-source_canonicalized_v4.0.1.xml"/>
```

musicEnfanthen committed:
fix(build.xml): rename default file for compared old version

bwbohl commented:
@musicEnfanthen could you resolve the tiny conflict, please?

musicEnfanthen committed:
Merge branch 'develop' into add-comparison

musicEnfanthen commented:
@bwbohl Done, thanks for the pointer.

musicEnfanthen committed:
Merge branch 'develop' into add-comparison

bwbohl committed:
check if old.file for comparison is available, otherwise don't start transformation and fail comparison

bwbohl commented on BUILD_COMMANDLINE.md:
```diff
@@ -152,5 +152,6 @@ The following targets can be called using `ant <target>`:
 | `build-guidelines-pdf` | Builds the PDF version of the MEI guidelines. (Calls `build-guidelines-html` before execution.) |
 | `init` | Initializes the build environment, i.e., checks if saxon and xerces are available, and if not, downloads jar files for Saxon, Xerces and adds them to the `lib` folder. Checks if prince is available. |
 | `init-mei-classpath` | Initializes the mei.classpath, which is essential for the schema generation, by prepending the jar files contained in the lib directory to the java classpath. Will be called automatically if needed. |
+| `compare-versions` | Compares the canonicalized sources of the MEI dev version with the previous stable version and creates an HTML output; custom versions and output folder can be set via `-Dsource`, `-Dold` and `-Doutput` input params. <br> NB: path of old file is relative to `comparison.xml`, while others are relative to `build.xml`. |
```
```suggestion
| `compare-versions` | Compares the canonicalized sources of the MEI dev version with the previous stable version and creates an HTML output; custom versions and output folder can be set via `-Dsource`, `-Dold` and `-Doutput` input params. <br> NB: path of old file is relative to `comparison.xsl`, while others are relative to `build.xml`. |
```

bwbohl replied:
Moreover, the paths could probably also be absolute.

bwbohl commented on build.xml:
```diff
@@ -390,6 +390,48 @@
+        </condition>
+        <echo>Using source.file: ${source.file}</echo>
+
+        <!-- set old file to compare with (relative to comparison.xml) -->
```
```suggestion
        <!-- set old file to compare with (relative to comparison.xsl) -->
```

bwbohl commented on build.xml:
```diff
@@ -390,6 +390,48 @@
+        <echo>Using source.file: ${source.file}</echo>
+
+        <!-- set old file to compare with (relative to comparison.xml) -->
+        <property name="old.default" value="mei-source_canonicalized_v4.0.1.xml"/>
```
Why do you want this relative to `comparison.xsl`? The ant file will treat such paths relative to itself. So in terms of, e.g. checking the availability of $old file, it would make more sense to make this relative to build.xml or silently add the folder to a check (cf. https://github.com/musicEnfanthen/music-encoding/pull/1)

musicEnfanthen replied:
It was relative to comparison.xsl because the XSLT did not find it otherwise. But I think it should work with absolute path.

bwbohl committed:
check if old.default is picked, then check if available, else download

bwbohl commented:
Generally, it works great if the source-files are in place. But I wonder whether the $old.default should probably be fetched to temp or build instead of next to compariosn.xsl?
Moreover, we could think about moving the output to the dist.folder if one runs dist. Which makes me wonder why not by default put the build result of the taret in the build folder?

bwbohl committed:
move version variables to top for global availability

bwbohl committed:
add major version numers as variables

bwbohl committed:
add IDs to table row reflecting their content-name

bwbohl committed:
add links from changed content, classes, attributes to their entry or guidelines

bwbohl committed:
make links appear in same color

bwbohl committed:
add carets before and after element names

bwbohl committed:
add some TODOs

bwbohl committed:
create floating head and make heading elements sticky top

musicEnfanthen committed:
fix(build): fix typo

Co-authored-by: Benjamin W. Bohl <b.w.bohl@gmail.com>

musicEnfanthen committed:
fix(build): fix typo

Co-authored-by: Benjamin W. Bohl <b.w.bohl@gmail.com>

musicEnfanthen commented:
> Moreover, we could think about moving the output to the dist.folder if one runs dist. Which makes me wonder why not by default put the build result of the taret in the build folder?

The comparions.html needs to be in the same folder as the resources folder to be displayed correctly (with css and js). That's the only reason why I moved the output file next to the XSLT.

musicEnfanthen committed:
Merge pull request #1 from bwbohl/origin/pr/1259-review

PR 1259 review additions

musicEnfanthen committed:
Merge pull request #2 from bwbohl/ftr/comparison-improvements

Comparison improvements

musicEnfanthen committed:
fix(build): use absolute path to default old file in build folder

musicEnfanthen committed:
fix(build): if not available get default old source from schema repo

musicEnfanthen committed:
fix(build): check if old source is available and fail early if not

musicEnfanthen committed:
fix(build): fix docs

musicEnfanthen committed:
Merge branch 'add-comparison' of github.com:musicEnfanthen/music-encoding into add-comparison

# Conflicts:
#	build.xml

musicEnfanthen commented on build.xml:
```diff
@@ -390,6 +390,62 @@
+        </condition>
+        <echo>Using old.file: ${old.file}</echo>
+
+        <!-- check if old.file is available -->
```
```suggestion
        <!-- check if old.file is available, otherwise fail early -->
```

musicEnfanthen committed:
fix(build): update comment

musicEnfanthen commented:
Thanks for the additional improvements @bwbohl . Looks great.

Reworked your suggestions for the ant target. As sugested, it now defines the location of the default old file in the build folder (since we run canonicalize-source beforehand the build folder will be created in any case and can be used). If it does not find a file there, it will download it from the schema repo. It will fail early now when the old source file cannot be found (which is probably only possible if there is a custom user input since downloaded default file should be available). So the checks if the file is available can be skipped with later steps. I left the output folder in the utils/compare_versions directory since the html file will need the resources to be displayed correctly. We can improve and move that in the future, but would keep it for now.

In the XSLT, I refactored the version variables and guideline links so that it works now with dev and major versions.

musicEnfanthen committed:
fix(build): simplify docs and use absolute path only for java call

musicEnfanthen committed:
fix(comparison): get version from defaultVal in meiversion

musicEnfanthen committed:
fix(comparison): get major version depending on dev

musicEnfanthen committed:
fix(comparison): resort variables

musicEnfanthen committed:
fix(comparison): adjust display of versions in html title and header

musicEnfanthen committed:
fix(comparison): simplify handling of guidelines links (working with dev now)

bwbohl commented on build.xml:
```diff
@@ -390,6 +390,62 @@
+            <arg value="-xsl:${basedir}/utils/compare_versions/comparison.xsl"/>
+            <arg value="-xi:on"/>
+            <arg value="output.folder=${output.folder}"/>
+            <arg value="old.version.filename=/${old.file}"/>
```
Of course you're right that an absolute path will make the transformation work, no matter, where `${old.source}` is located. Running `ant compare-versions` natively from the music-encoding folder fails though. Moreover, setting it absolute here will prevent submitting an arbitrary file path. 

```suggestion
            <arg value="old.version.filename=${old.file}"/>
```

The absolute path for `${old.default}` could be achieved by prepending `${basedir}` in its definition (see above).

bwbohl replied:
```suggestion
            <arg value="old.version.filename=file:${old.file}"/>
```

bwbohl commented on build.xml:
```diff
@@ -390,6 +390,62 @@
+        <echo>Using source.file: ${source.file}</echo>
+
+        <!-- set old file to compare with -->
+        <property name="old.default" value="${dir.build}/mei-source_canonicalized_v4.0.1.xml"/>
```
```suggestion
        <property name="old.default" value="${basedir}/${dir.build}/mei-source_canonicalized_v4.0.1.xml"/>
```

see comment below

bwbohl commented:
with the two mentioned changes it works fine locally.

musicEnfanthen commented:
Unfortunately, your suggestions will not work on Win. The XSLT does not find the old.file then.

bwbohl commented:
In the current state, it does not work on macOS, though…

musicEnfanthen committed:
fix(build): fix path issue

Now working on mac and win

Co-authored-by: Benjamin W. Bohl <b.w.bohl@gmail.com>
